src/ParseCallMap.hs

```diff
@@ -0,0 +1,313 @@
+-- calltreeForm         ::= ( calltree <subcallForm>* )
+-- subcallForm          ::= ( call <identForm> )
+--                       |  ( choice <calltreeForm> <calltreeForm> )
+--                       |  ( seq <subcallForm>* )
+-- identForm            ::= ( ident <string_lit> <nodeForm> )
+-- nodeForm             ::= noNode
+--                       |  ( node <cspanForm> )
+-- cspanForm            ::= noPos
+--                       |  builtin
+--                       |  internal
+--                       |  ( span <sourceForm> <int_lit> <sourceForm> )
+-- sourceForm           ::= ( source <string_lit> <int_lit> <int_lit> <int_lit> )
+--
+
+
+-- | An identifier
+newtype Sidentifier = Sidentifier { sidentText :: Text.Text }
+  deriving (Show, Eq, IsString)
+
+-- | Parse the given callmap graph file.  Assumes UTF8 encoding (will use
+-- U+FFFD replacement character on UTF8 decoding errors).
+fromFile :: FilePath -> IO (Either CallMapParseError CallMap)
+fromFile path = do
+  bs <- BSL.readFile path
+  let txt = TextL.decodeUtf8With TextL.lenientDecode bs
+  return (fromSource path txt)
+
+
+-- | @fromSource path txt@ parses @txt@ and returns the resulting 'CallMap'.
+-- (The @path@ is used in error messages)
+fromSource :: FilePath -> TextL.Text -> Either CallMapParseError CallMap
+fromSource = parse parser
+
+parser :: Parser CallMap
+parser = between ws eof callmapForm
+
+callmapForm :: Parser CallMap
+callmapForm = theForm "callmap" (Map.fromList <$> many functionForm)
+
+type CallMapItem = (NodeInfo, CallTree Ident, PermissionActionSet)
+
+functionForm :: Parser (Ident, CallMapItem)
+functionForm = theForm "function" ((,)
+                                    <$> identForm
+                                    <*> (mkItem <$> nodeForm <*> actionsForm <*> calltreeForm))
+  where
+    mkItem node actions tree = (node, tree, actions)
+
+actionsForm :: Parser PermissionActionSet
+actionsForm = theForm "actions" (HashSet.fromList <$> many permissionActionForm)
+
+permissionActionForm :: Parser PermissionAction
+permissionActionForm =
+  aForm (keyword <*> (permissionName <$> identifierLiteral))
+  where
+    keyword = asum (map kw actions)
+    kw :: (Sidentifier, (PermissionName -> PermissionAction)) -> Parser (PermissionName -> PermissionAction)
+    kw (ident,f) = f <$ theIdentifier ident
+    actions :: [(Sidentifier, PermissionName -> PermissionAction)]
+    actions = [ ("need", Need)
+              , ("use", Use)
+              , ("grant", Grant)
+              , ("revoke", Revoke)
+              , ("deny", Deny)
+              , ("waive", Waive)
+              ]
+    permissionName :: Spanning Sidentifier -> PermissionName
+    permissionName = PermissionName . sidentText . extractSpan
+
+calltreeForm :: Parser (CallTree Ident)
+calltreeForm = theForm "calltree" seqForest
+
+seqForest :: Parser (CallTree Ident)
+seqForest = joinForest <$> many subcallForm
+  where
+    joinForest :: [CallTree a] -> CallTree a
```

Comment:
maybe.  it's really the `mconcat` of the `Sequence`/`Nop` monoid structure for `CallTree`. However I didn't feel great about writing the `Monoid` instance because I couldn't convince myself that the `Choice`/`Nop` monoid is clearly less useful.  In any case I was going to spend tomorrow playing around with a sequence-of-choices (ie two data types) presentation of call trees to see if I liked it better in `Check.Permissions` in which case the monoid structure would be more natural and this function would go away.